# Pokémon API Data Engineering & Analytics Pipeline

In [38]:
# Import the libraries we will use for data analysis, numerical calculations, and visualisation.
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

print("Libraries imported successfully!")

Libraries imported successfully!


In [39]:
# Load the raw Pokémon CSV file into a Pandas DataFrame so we can analyse the data.
df = pd.read_csv("pokemon_raw.csv")
print("Dataset loaded successfully!")

Dataset loaded successfully!


In [40]:
# Display the first 5 rows of the dataset to get our first look at the data.
df.head()

,id,name,height,weight,base_experience,types,abilities,hp,attack,defense,special_attack,special_defense,speed
0,1,bulbasaur,7,69,64.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",45,49,49,65,65,45
1,2,ivysaur,10,130,142.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",60,62,63,80,80,60
2,3,venusaur,20,1000,236.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",80,82,83,100,100,80
3,4,charmander,6,85,62.0,['fire'],"['blaze', 'solar-power']",39,52,43,60,50,65
4,5,charmeleon,11,190,142.0,['fire'],"['blaze', 'solar-power']",58,64,58,80,65,80


In [41]:
# Display the first 10 rows so we can inspect more records.
df.head(10)

,id,name,height,weight,base_experience,types,abilities,hp,attack,defense,special_attack,special_defense,speed
0,1,bulbasaur,7,69,64.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",45,49,49,65,65,45
1,2,ivysaur,10,130,142.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",60,62,63,80,80,60
2,3,venusaur,20,1000,236.0,"['grass', 'poison']","['overgrow', 'chlorophyll']",80,82,83,100,100,80
3,4,charmander,6,85,62.0,['fire'],"['blaze', 'solar-power']",39,52,43,60,50,65
4,5,charmeleon,11,190,142.0,['fire'],"['blaze', 'solar-power']",58,64,58,80,65,80
5,6,charizard,17,905,240.0,"['fire', 'flying']","['blaze', 'solar-power']",78,84,78,109,85,100
6,7,squirtle,5,90,63.0,['water'],"['torrent', 'rain-dish']",44,48,65,50,64,43
7,8,wartortle,10,225,142.0,['water'],"['torrent', 'rain-dish']",59,63,80,65,80,58
8,9,blastoise,16,855,239.0,['water'],"['torrent', 'rain-dish']",79,83,100,85,105,78
9,10,caterpie,3,29,39.0,['bug'],"['shield-dust', 'run-away']",45,30,35,20,20,45


In [42]:
# Display the last 5 rows to check how the dataset ends.
df.tail()

,id,name,height,weight,base_experience,types,abilities,hp,attack,defense,special_attack,special_defense,speed
1346,10322,tatsugiri-curly-mega,6,240,NaN,"['dragon', 'water']",[],68,65,90,135,125,92
1347,10323,tatsugiri-droopy-mega,6,240,NaN,"['dragon', 'water']",[],68,65,90,135,125,92
1348,10324,tatsugiri-stretchy-mega,6,240,NaN,"['dragon', 'water']",[],68,65,90,135,125,92
1349,10325,baxcalibur-mega,21,3150,NaN,"['dragon', 'ice']",[],115,175,117,105,101,87
1350,10326,meowstic-female-mega,8,101,NaN,['psychic'],['trace'],74,48,76,143,101,124


In [43]:
# Show the number of rows and columns in the dataset.
df.shape

(1351, 13)

In [44]:
# Display all column names so we know which variables are available for analysis.
df.columns

Index(['id', 'name', 'height', 'weight', 'base_experience', 'types',
       'abilities', 'hp', 'attack', 'defense', 'special_attack',
       'special_defense', 'speed'],
      dtype='object')

In [45]:
# Display detailed information about the DataFrame, including data types and non-missing values.
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1351 entries, 0 to 1350
Data columns (total 13 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1351 non-null   int64  
 1   name             1351 non-null   object 
 2   height           1351 non-null   int64  
 3   weight           1351 non-null   int64  
 4   base_experience  1302 non-null   float64
 5   types            1351 non-null   object 
 6   abilities        1351 non-null   object 
 7   hp               1351 non-null   int64  
 8   attack           1351 non-null   int64  
 9   defense          1351 non-null   int64  
 10  special_attack   1351 non-null   int64  
 11  special_defense  1351 non-null   int64  
 12  speed            1351 non-null   int64  
dtypes: float64(1), int64(9), object(3)
memory usage: 137.3+ KB


In [46]:
# Count the number of missing values (NaN) in every column.
df.isnull().sum()

id                  0
name                0
height              0
weight              0
base_experience    49
types               0
abilities           0
hp                  0
attack              0
defense             0
special_attack      0
special_defense     0
speed               0
dtype: int64

In [47]:
# Find the Pokémon records where base_experience is missing (NaN).
df[df["base_experience"].isnull()]

,id,name,height,weight,base_experience,types,abilities,hp,attack,defense,special_attack,special_defense,speed
1302,10278,clefable-mega,17,423,NaN,"['fairy', 'flying']",['magic-bounce'],95,80,93,135,110,70
1303,10279,victreebel-mega,45,1255,NaN,"['grass', 'poison']",['innards-out'],80,125,85,135,95,70
1304,10280,starmie-mega,23,800,NaN,"['water', 'psychic']",['huge-power'],60,100,105,130,105,120
1305,10281,dragonite-mega,22,2900,NaN,"['dragon', 'flying']",['multiscale'],91,124,115,145,125,100
1306,10282,meganium-mega,24,2010,NaN,"['grass', 'fairy']",['mega-sol'],80,92,115,143,115,80
1307,10283,feraligatr-mega,23,1088,NaN,"['water', 'dragon']",['dragonize'],85,160,125,89,93,78
1308,10284,skarmory-mega,17,404,NaN,"['steel', 'flying']",['stalwart'],65,140,110,40,100,110
1309,10285,froslass-mega,26,296,NaN,"['ice', 'ghost']",['snow-warning'],70,80,70,140,100,120
1310,10286,emboar-mega,18,1803,NaN,"['fire', 'fighting']",['mold-breaker'],110,148,75,110,110,75
1311,10287,excadrill-mega,9,600,NaN,"['ground', 'steel']",['piercing-drill'],110,165,100,65,65,103


In [48]:
# Show the first and last Pokémon that have missing base_experience values.
# This helps us confirm whether the missing values are concentrated in a particular group.
missing_exp = df[df["base_experience"].isna()]

print("Number of missing values:", len(missing_exp))
print("\nFirst missing record:")
print(missing_exp[["id", "name", "base_experience"]].head())

print("\nLast missing record:")
print(missing_exp[["id", "name", "base_experience"]].tail())

Number of missing values: 49

First missing record:
         id             name  base_experience
1302  10278    clefable-mega              NaN
1303  10279  victreebel-mega              NaN
1304  10280     starmie-mega              NaN
1305  10281   dragonite-mega              NaN
1306  10282    meganium-mega              NaN

Last missing record:
         id                     name  base_experience
1346  10322     tatsugiri-curly-mega              NaN
1347  10323    tatsugiri-droopy-mega              NaN
1348  10324  tatsugiri-stretchy-mega              NaN
1349  10325          baxcalibur-mega              NaN
1350  10326     meowstic-female-mega              NaN


In [49]:
# Find the first row where base_experience is missing.
# We want to determine whether the missing values start at a particular point in the dataset.
first_missing_index = df["base_experience"].isna().idxmax()

print("First missing DataFrame index:", first_missing_index)
print(df.loc[first_missing_index, ["id", "name", "base_experience"]])

First missing DataFrame index: 1302
id                         10278
name               clefable-mega
base_experience              NaN
Name: 1302, dtype: object


In [50]:
# Display the Pokémon around the first missing value.
# We are comparing the normal records immediately before the missing data
# with the records where base_experience becomes NaN.
df.loc[1295:1310, ["id", "name", "base_experience"]]

,id,name,base_experience
1295,10271,miraidon-glide-mode,335.0
1296,10272,ursaluna-bloodmoon,275.0
1297,10273,ogerpon-wellspring-mask,275.0
1298,10274,ogerpon-hearthflame-mask,275.0
1299,10275,ogerpon-cornerstone-mask,275.0
1300,10276,terapagos-terastal,90.0
1301,10277,terapagos-stellar,90.0
1302,10278,clefable-mega,NaN
1303,10279,victreebel-mega,NaN
1304,10280,starmie-mega,NaN


In [51]:
# Display the last 60 Pokémon and their base_experience values.
# This lets us inspect the transition from populated values to NaN.
df[["id", "name", "base_experience"]].tail(60)

,id,name,base_experience
1291,10267,koraidon-gliding-build,335.0
1292,10268,miraidon-low-power-mode,335.0
1293,10269,miraidon-drive-mode,335.0
1294,10270,miraidon-aquatic-mode,335.0
1295,10271,miraidon-glide-mode,335.0
1296,10272,ursaluna-bloodmoon,275.0
1297,10273,ogerpon-wellspring-mask,275.0
1298,10274,ogerpon-hearthflame-mask,275.0
1299,10275,ogerpon-cornerstone-mask,275.0
1300,10276,terapagos-terastal,90.0


In [52]:
# Check the data type Pandas assigned to the base_experience column.
# Because the column contains numbers and NaN values, this tells us how Pandas represents them.
df["base_experience"].dtype

dtype('float64')

In [53]:
# Count the number of non-missing base_experience values.
# This gives us the number of Pokémon that actually have a value in this column.
df["base_experience"].count()

np.int64(1302)

In [54]:
# Calculate what percentage of the base_experience column contains missing values.
# This helps us measure how significant the missing-data problem is.
missing_percentage = df["base_experience"].isna().mean() * 100

print(f"Missing base_experience: {missing_percentage:.2f}%")

Missing base_experience: 3.63%


In [55]:
# Find the first and last DataFrame indexes where base_experience is missing.
# This helps us see whether the missing values are scattered or form one continuous block.
missing_indexes = df.index[df["base_experience"].isna()]

print("First missing index:", missing_indexes.min())
print("Last missing index:", missing_indexes.max())
print("Number of missing values:", len(missing_indexes))

First missing index: 1302
Last missing index: 1350
Number of missing values: 49


In [56]:
# Display the Pokémon IDs and names for all rows where base_experience is missing.
# We want to identify whether these records have something in common.
df.loc[df["base_experience"].isna(), ["id", "name", "base_experience"]]

,id,name,base_experience
1302,10278,clefable-mega,NaN
1303,10279,victreebel-mega,NaN
1304,10280,starmie-mega,NaN
1305,10281,dragonite-mega,NaN
1306,10282,meganium-mega,NaN
1307,10283,feraligatr-mega,NaN
1308,10284,skarmory-mega,NaN
1309,10285,froslass-mega,NaN
1310,10286,emboar-mega,NaN
1311,10287,excadrill-mega,NaN


In [57]:
# Request one Pokémon directly from PokéAPI.
# We are checking whether the missing base_experience came from the API itself
# rather than being caused by our CSV or Pandas transformation.
import requests

url = "https://pokeapi.co/api/v2/pokemon/clefable-mega"

response = requests.get(url, timeout=20)

print("Status code:", response.status_code)

pokemon = response.json()

print("Name:", pokemon["name"])
print("Base experience:", pokemon["base_experience"])

Status code: 200
Name: clefable-mega
Base experience: None


In [58]:
# Calculate the percentage of missing values in every column.
# This helps us identify which columns have data-quality issues
# and how large each missing-value problem is.
missing_percentage = df.isna().mean() * 100

print(missing_percentage)

id                 0.000000
name               0.000000
height             0.000000
weight             0.000000
base_experience    3.626943
types              0.000000
abilities          0.000000
hp                 0.000000
attack             0.000000
defense            0.000000
special_attack     0.000000
special_defense    0.000000
speed              0.000000
dtype: float64


In [59]:
df["base_experience"].dtype

dtype('float64')

In [60]:
# Check how many completely duplicated rows exist in the dataset.
# A duplicate row means every value in that row is identical to another row.
df.duplicated().sum()

np.int64(0)

In [61]:
# Check whether each row is an exact duplicate of a previous row.
df.duplicated()


0       False
1       False
2       False
3       False
4       False
        ...  
1346    False
1347    False
1348    False
1349    False
1350    False
Length: 1351, dtype: bool

In [62]:
# Count how many duplicate Pokémon IDs exist.
# We use sum() because True counts as 1 and False counts as 0.
df["id"].duplicated().sum()

np.int64(0)

In [63]:
# Generate descriptive statistics for the numerical columns.
# This helps us understand the distribution, typical values,
# minimums, maximums, and spread of the Pokémon statistics.
df.describe()

,id,height,weight,base_experience,hp,attack,defense,special_attack,special_defense,speed
count,1351.000000,1351.000000,1351.000000,1302.000000,1351.000000,1351.000000,1351.000000,1351.000000,1351.000000,1351.000000
mean,2841.692080,20.443375,991.059956,161.943932,71.673575,82.757957,76.193930,75.418209,74.115470,72.000000
std,4139.019112,53.238467,1953.115812,82.532554,26.827529,32.351906,30.712902,34.026272,27.800826,30.642043
min,1.000000,1.000000,0.000000,36.000000,1.000000,5.000000,5.000000,10.000000,20.000000,5.000000
25%,338.500000,6.000000,94.500000,71.000000,55.000000,60.000000,55.000000,50.000000,54.000000,49.000000
50%,676.000000,11.000000,326.000000,165.000000,70.000000,80.000000,70.000000,69.000000,70.000000,70.000000
75%,1013.500000,17.000000,853.500000,222.500000,85.000000,102.500000,95.000000,97.000000,90.000000,95.000000
max,10326.000000,1000.000000,10000.000000,608.000000,255.000000,190.000000,250.000000,216.000000,250.000000,200.000000


## Now we have checked the following 
✅ Shape
✅ Columns
✅ Data types
✅ Missing values
✅ Duplicate rows
✅ Duplicate IDs
✅ Descriptive statistics

In [64]:
# Display the data type of every column in the DataFrame.
# This helps us understand how Pandas is storing each field before we transform the data.
df.dtypes

id                   int64
name                object
height               int64
weight               int64
base_experience    float64
types               object
abilities           object
hp                   int64
attack               int64
defense              int64
special_attack       int64
special_defense      int64
speed                int64
dtype: object

In [65]:
# Count how many different type combinations exist in the dataset.
# This helps us understand the structure of the types column before transforming it.
df["types"].value_counts()

types
['normal']              86
['water']               84
['electric']            56
['grass']               49
['psychic']             49
                        ..
['psychic', 'ice']       1
['ice', 'rock']          1
['fighting', 'fire']     1
['grass', 'water']       1
['grass', 'rock']        1
Name: count, Length: 221, dtype: int64

In [66]:
# Check the Python data type of the first value in the types column.
# This tells us whether Pandas is storing the value as a real Python list or as text.
type(df["types"].iloc[0])

str

In [67]:
# Display the first five values from the types column.
# We want to see exactly how the list data is currently stored as strings.
print(df["types"].head())

0    ['grass', 'poison']
1    ['grass', 'poison']
2    ['grass', 'poison']
3               ['fire']
4               ['fire']
Name: types, dtype: object


In [68]:
# Display the first five values from the abilities column.
# We want to see whether abilities has the same string representation as types.
print(df["abilities"].head())

0    ['overgrow', 'chlorophyll']
1    ['overgrow', 'chlorophyll']
2    ['overgrow', 'chlorophyll']
3       ['blaze', 'solar-power']
4       ['blaze', 'solar-power']
Name: abilities, dtype: object


In [70]:
# Import Python's Abstract Syntax Tree module.
# We will use ast.literal_eval() to safely convert list-like strings back into Python lists.
import ast

In [71]:
# Convert a string that looks like a Python list into an actual Python list.
ast.literal_eval("['grass', 'poison']")

['grass', 'poison']

In [72]:
# Check the Python data type of the converted value.
# We want to confirm that the string has been converted into a real list.
type(ast.literal_eval("['grass', 'poison']"))

list

In [73]:
# Convert every value in the types column from a string representation
# of a Python list into an actual Python list.
df["types"] = df["types"].apply(ast.literal_eval)

In [74]:
# Check the Python data type of the first value after transformation.
# We expect Pandas to now contain an actual Python list.
type(df["types"].iloc[0])

list

In [75]:
# Display the first five transformed values so we can visually verify the result.
print(df["types"].head())

0    [grass, poison]
1    [grass, poison]
2    [grass, poison]
3             [fire]
4             [fire]
Name: types, dtype: object


In [76]:
# Confirm that the first value in the types column is now a Python list.
# This verifies that our transformation was successful.
type(df["types"].iloc[0])

list

In [77]:
# Convert every value in the abilities column from a string representation
# of a Python list into an actual Python list.
df["abilities"] = df["abilities"].apply(ast.literal_eval)

In [79]:
# Check that the first value in the abilities column is now a Python list.
type(df["abilities"].iloc[0])

list

In [80]:
# Display the first five transformed abilities values.
# This lets us visually confirm that the data now contains lists.
print(df["abilities"].head())

0    [overgrow, chlorophyll]
1    [overgrow, chlorophyll]
2    [overgrow, chlorophyll]
3       [blaze, solar-power]
4       [blaze, solar-power]
Name: abilities, dtype: object


In [81]:
# Check the data type of the first value in both transformed columns.
# We want to confirm that both columns now contain Python lists.
print(type(df["types"].iloc[0]))
print(type(df["abilities"].iloc[0]))

<class 'list'>
<class 'list'>


In [82]:
# Count how many Pokémon types are stored for each Pokémon.
# This helps us understand whether Pokémon have one type, two types, or more.
df["types"].apply(len).value_counts()


types
2    765
1    586
Name: count, dtype: int64

In [84]:
# Create a new column showing how many types each Pokémon has.
df["type_count"] = df["types"].apply(len)

In [85]:
# Display the Pokémon name, their types, and the number of types.
df[["name", "types", "type_count"]].head(10)

,name,types,type_count
0,bulbasaur,"[grass, poison]",2
1,ivysaur,"[grass, poison]",2
2,venusaur,"[grass, poison]",2
3,charmander,[fire],1
4,charmeleon,[fire],1
5,charizard,"[fire, flying]",2
6,squirtle,[water],1
7,wartortle,[water],1
8,blastoise,[water],1
9,caterpie,[bug],1


In [86]:
# Count how many abilities are stored for each Pokémon.
# This helps us understand the structure of the abilities column.
df["abilities"].apply(len).value_counts()

abilities
3    613
2    375
1    352
0     11
Name: count, dtype: int64

In [87]:
# Find Pokémon that have an empty abilities list.
# We want to inspect these records before deciding whether any transformation is needed.
df.loc[df["abilities"].apply(len) == 0, ["id", "name", "abilities"]]

,id,name,abilities
1325,10301,zygarde-mega,[]
1335,10311,heatran-mega,[]
1336,10312,darkrai-mega,[]
1340,10316,golisopod-mega,[]
1341,10317,magearna-mega,[]
1342,10318,magearna-original-mega,[]
1343,10319,zeraora-mega,[]
1346,10322,tatsugiri-curly-mega,[]
1347,10323,tatsugiri-droopy-mega,[]
1348,10324,tatsugiri-stretchy-mega,[]


In [89]:
# Create a new column showing how many abilities each Pokémon has.
df["ability_count"] = df["abilities"].apply(len)

In [90]:
# Display Pokémon names, their abilities, and the number of abilities.
df[["name", "abilities", "ability_count"]].head(10)

,name,abilities,ability_count
0,bulbasaur,"[overgrow, chlorophyll]",2
1,ivysaur,"[overgrow, chlorophyll]",2
2,venusaur,"[overgrow, chlorophyll]",2
3,charmander,"[blaze, solar-power]",2
4,charmeleon,"[blaze, solar-power]",2
5,charizard,"[blaze, solar-power]",2
6,squirtle,"[torrent, rain-dish]",2
7,wartortle,"[torrent, rain-dish]",2
8,blastoise,"[torrent, rain-dish]",2
9,caterpie,"[shield-dust, run-away]",2


In [91]:
# Check the DataFrame shape after creating the two new derived columns.
df.shape

(1351, 15)

In [92]:
# Display the last few rows of the DataFrame to verify our new derived columns.
# We want to confirm that type_count and ability_count were added correctly.
df.tail()


,id,name,height,weight,base_experience,types,abilities,hp,attack,defense,special_attack,special_defense,speed,type_count,ability_count
1346,10322,tatsugiri-curly-mega,6,240,NaN,"[dragon, water]",[],68,65,90,135,125,92,2,0
1347,10323,tatsugiri-droopy-mega,6,240,NaN,"[dragon, water]",[],68,65,90,135,125,92,2,0
1348,10324,tatsugiri-stretchy-mega,6,240,NaN,"[dragon, water]",[],68,65,90,135,125,92,2,0
1349,10325,baxcalibur-mega,21,3150,NaN,"[dragon, ice]",[],115,175,117,105,101,87,2,0
1350,10326,meowstic-female-mega,8,101,NaN,[psychic],[trace],74,48,76,143,101,124,1,1


In [93]:
# Display each column name together with its Pandas data type.
# This gives us a clean overview of how our transformed dataset is currently structured.
df.dtypes

id                   int64
name                object
height               int64
weight               int64
base_experience    float64
types               object
abilities           object
hp                   int64
attack               int64
defense              int64
special_attack       int64
special_defense      int64
speed                int64
type_count           int64
ability_count        int64
dtype: object

In [94]:
# Check the Python data type of values inside the name, types, and abilities columns.
# This confirms exactly what kind of objects our DataFrame is storing in these columns.
print(type(df["name"].iloc[0]))
print(type(df["types"].iloc[0]))
print(type(df["abilities"].iloc[0]))

<class 'str'>
<class 'list'>
<class 'list'>


In [95]:
# Count the number of unique Pokémon names.
# This helps us check whether the name column contains duplicate Pokémon records.
df["name"].nunique()

1351

In [96]:
# Expand the types list so each individual Pokémon type becomes a separate row.
# This allows us to count individual types instead of counting entire type combinations.
type_counts = df["types"].explode().value_counts()
type_counts

types
water       192
normal      160
grass       156
flying      154
psychic     142
dragon      117
electric    114
fire        109
fighting    109
poison      106
bug         106
rock        104
steel        99
dark         99
ghost        96
ground       96
fairy        88
ice          69
Name: count, dtype: int64

In [97]:
# Count the total number of individual type assignments in the dataset.
# This should equal one type for every single-type Pokémon plus two types for every dual-type Pokémon.
type_counts.sum()

np.int64(2116)

In [98]:
# Expand the abilities list so each individual ability becomes a separate row.
# This allows us to count how frequently each ability appears across the Pokémon dataset.
ability_counts = df["abilities"].explode().value_counts()
ability_counts

abilities
swift-swim        48
sturdy            48
intimidate        47
levitate          45
keen-eye          43
                  ..
piercing-drill     1
eelevate           1
fire-mane          1
aura-guard         1
spicy-spray        1
Name: count, Length: 313, dtype: int64

In [99]:
# Check whether exploding the abilities column created missing values.
# This helps us understand how empty ability lists are represented after the transformation.
ability_counts.isna().sum()

np.int64(0)

In [100]:
# Expand the abilities list into individual rows and count the missing values created by empty lists.
# This lets us see what happens to the 11 Pokémon that originally had abilities = [].
exploded_abilities = df["abilities"].explode()
exploded_abilities.isna().sum()

np.int64(11)

In [101]:
# Find the Pokémon whose abilities list is empty.
# We want to inspect the actual records before deciding how to handle these values.
df.loc[df["abilities"].apply(len) == 0, ["id", "name", "abilities"]]

,id,name,abilities
1325,10301,zygarde-mega,[]
1335,10311,heatran-mega,[]
1336,10312,darkrai-mega,[]
1340,10316,golisopod-mega,[]
1341,10317,magearna-mega,[]
1342,10318,magearna-original-mega,[]
1343,10319,zeraora-mega,[]
1346,10322,tatsugiri-curly-mega,[]
1347,10323,tatsugiri-droopy-mega,[]
1348,10324,tatsugiri-stretchy-mega,[]


In [102]:
# Calculate the percentage of Pokémon whose abilities list is empty.
# This tells us how significant the empty-ability records are in the dataset.
empty_abilities_percentage = (df["abilities"].apply(len) == 0).mean() * 100
print(f"Pokémon with empty abilities: {empty_abilities_percentage:.2f}%")

Pokémon with empty abilities: 0.81%


## Profile so far 
Rows                    → 1,351
Columns                 → 15
Duplicate rows          → 0
Duplicate IDs           → 0
Duplicate names         → 0
Missing base_experience → 49 (3.63%)
Empty abilities        → 11 (0.81%)
Single-type Pokémon     → 586
Dual-type Pokémon       → 765
Type assignments        → 2,116

In [103]:
# Find the 10 Pokémon with the greatest height values.
# We want to investigate whether the extreme values are legitimate or potential data-quality issues.
df[["name", "height"]].sort_values("height", ascending=False).head(10)

,name,height
1214,eternatus-eternamax,1000
1244,centiskorch-gmax,750
1249,duraludon-gmax,430
1239,coalossal-gmax,420
1235,inteleon-gmax,400
1230,snorlax-gmax,350
1224,meowth-gmax,330
1246,grimmsnarl-gmax,320
1247,alcremie-gmax,300
1250,urshifu-single-strike-gmax,290


In [104]:
# Find the 10 Pokémon with the greatest weight values.
# We want to investigate whether the extreme weight values follow a similar pattern to height.
df[["name", "weight"]].sort_values("weight", ascending=False).head(10)

,name,weight
1249,duraludon-gmax,10000
1251,urshifu-rapid-strike-gmax,10000
1252,toxtricity-low-key-gmax,10000
1244,centiskorch-gmax,10000
1243,toxtricity-amped-gmax,10000
1241,appletun-gmax,10000
1250,urshifu-single-strike-gmax,10000
1247,alcremie-gmax,10000
1246,grimmsnarl-gmax,10000
1245,hatterene-gmax,10000


In [105]:
# Find the 10 Pokémon with the highest HP values.
# We want to investigate the range of the Pokémon battle statistics and identify extreme values.
df[["name", "hp"]].sort_values("hp", ascending=False).head(10)

,name,hp
1214,eternatus-eternamax,255
241,blissey,255
112,chansey,250
798,guzzlord,223
1144,zygarde-complete,216
1325,zygarde-mega,216
894,regidrago,200
201,wobbuffet,190
320,wailord,170
974,cetitan,170


In [106]:
# Find the 10 Pokémon with the highest attack values.
# We want to identify extreme attack values and determine whether they appear to be legitimate source data.
df[["name", "attack"]].sort_values("attack", ascending=False).head(10)

,name,attack
1067,mewtwo-mega-x,190
1071,heracross-mega,185
797,kartana,181
1102,groudon-primal,180
1103,rayquaza-mega,180
1025,deoxys-attack,180
1349,baxcalibur-mega,175
1082,garchomp-mega,170
1046,kyurem-black,170
1181,necrozma-ultra,167


In [107]:
# Find the 10 Pokémon with the highest defense values.
# We want to identify extreme defense values and determine whether they appear to be legitimate source data.
df[["name", "defense"]].sort_values("defense", ascending=False).head(10)


,name,defense
1214,eternatus-eternamax,250
1096,steelix-mega,230
1077,aggron-mega,230
212,shuckle,230
804,stakataka,211
207,steelix,200
376,regirock,200
1267,avalugg-hisui,184
712,avalugg,184
1095,slowbro-mega,180


In [108]:
# Find the 10 Pokémon with the highest special_attack values.
# We want to identify extreme special-attack values and determine whether they appear to be legitimate source data.
df[["name", "special_attack"]].sort_values("special_attack", ascending=False).head(10)

,name,special_attack
1325,zygarde-mega,216
1068,mewtwo-mega-y,194
1025,deoxys-attack,180
1103,rayquaza-mega,180
1101,kyogre-primal,180
1335,heatran-mega,175
1315,chandelure-mega,175
1061,alakazam-mega,175
795,xurkitree,173
1047,kyurem-white,170


In [109]:
# Find the 10 Pokémon with the highest special_defense values.
# We want to identify extreme special-defense values and determine whether they appear to be legitimate source data.
df[["name", "special_defense"]].sort_values("special_defense", ascending=False).head(10)

,name,special_defense
1214,eternatus-eternamax,250
212,shuckle,230
377,regice,200
1323,dragalge-mega,163
1026,deoxys-defense,160
1101,kyogre-primal,160
670,florges,154
248,lugia,154
249,ho-oh,154
1266,goodra-hisui,150


In [111]:
# Find the 10 Pokémon with the highest speed values.
# We want to identify extreme speed values and determine whether they appear to be legitimate source data.
df[["name", "speed"]].sort_values("speed", ascending=False).head(10)

,name,speed
893,regieleki,200
1027,deoxys-speed,180
290,ninjask,160
1343,zeraora-mega,153
1331,absol-mega-z,151
1334,lucario-mega-z,151
1333,garchomp-mega-z,151
794,pheromosa,151
1256,electrode-hisui,150
100,electrode,150


In [113]:
# Display all Pokémon records where base_experience is missing.
# We inspect these records before deciding how the missing values should be handled.
missing_base_experience = df[df["base_experience"].isna()]
missing_base_experience[["id", "name", "base_experience"]]

,id,name,base_experience
1302,10278,clefable-mega,NaN
1303,10279,victreebel-mega,NaN
1304,10280,starmie-mega,NaN
1305,10281,dragonite-mega,NaN
1306,10282,meganium-mega,NaN
1307,10283,feraligatr-mega,NaN
1308,10284,skarmory-mega,NaN
1309,10285,froslass-mega,NaN
1310,10286,emboar-mega,NaN
1311,10287,excadrill-mega,NaN


In [114]:
# Calculate the percentage of Pokémon records with and without base_experience.
# This helps us understand how complete the source field is before making our final cleaning decision.
base_experience_missing = df["base_experience"].isna().sum()
base_experience_present = df["base_experience"].notna().sum()
total_records = len(df)
print("Base experience present:", base_experience_present)
print("Base experience missing:", base_experience_missing)
print("Total records:", total_records)
print(f"Missing percentage: {(base_experience_missing / total_records) * 100:.2f}%")

Base experience present: 1302
Base experience missing: 49
Total records: 1351
Missing percentage: 3.63%


In [115]:
# Document the decision for handling missing base_experience values.
# The source API does not provide these values, so we preserve them as missing instead of inventing data.
base_experience_missing = df["base_experience"].isna().sum()
print(f"Missing base_experience values: {base_experience_missing}")
print("Decision: Preserve missing values as NaN.")

Missing base_experience values: 49
Decision: Preserve missing values as NaN.


In [116]:
# Display Pokémon records where the abilities list is empty.
# We inspect these records before deciding how empty ability collections should be handled.
empty_abilities = df[df["ability_count"] == 0]
empty_abilities[["id", "name", "abilities", "ability_count"]]

,id,name,abilities,ability_count
1325,10301,zygarde-mega,[],0
1335,10311,heatran-mega,[],0
1336,10312,darkrai-mega,[],0
1340,10316,golisopod-mega,[],0
1341,10317,magearna-mega,[],0
1342,10318,magearna-original-mega,[],0
1343,10319,zeraora-mega,[],0
1346,10322,tatsugiri-curly-mega,[],0
1347,10323,tatsugiri-droopy-mega,[],0
1348,10324,tatsugiri-stretchy-mega,[],0


In [117]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1351 entries, 0 to 1350
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   id               1351 non-null   int64  
 1   name             1351 non-null   object 
 2   height           1351 non-null   int64  
 3   weight           1351 non-null   int64  
 4   base_experience  1302 non-null   float64
 5   types            1351 non-null   object 
 6   abilities        1351 non-null   object 
 7   hp               1351 non-null   int64  
 8   attack           1351 non-null   int64  
 9   defense          1351 non-null   int64  
 10  special_attack   1351 non-null   int64  
 11  special_defense  1351 non-null   int64  
 12  speed            1351 non-null   int64  
 13  type_count       1351 non-null   int64  
 14  ability_count    1351 non-null   int64  
dtypes: float64(1), int64(11), object(3)
memory usage: 158.4+ KB


In [ ]:
# Calculate the percentage of Pokémon records with and without base_experience.
# This helps us understand how complete the source field is before making our final cleaning decision.
base_experience_missing = df["base_experience"].isna().sum()
base_experience_present = df["base_experience"].notna().sum()
total_records = len(df)
print("Base experience present:", base_experience_present)
print("Base experience missing:", base_experience_missing)
print("Total records:", total_records)
print(f"Missing percentage: {(base_experience_missing / total_records) * 100:.2f}%")

Base experience present: 1302
Base experience missing: 49
Total records: 1351
Missing percentage: 3.63%


In [118]:
# Check one Pokémon directly from PokéAPI to verify whether the empty abilities list comes from the source API.
import requests
url = "https://pokeapi.co/api/v2/pokemon/zygarde-mega"
response = requests.get(url, timeout=20)
print("Status code:", response.status_code)
pokemon = response.json()
print("Name:", pokemon["name"])
print("Abilities:", pokemon["abilities"])

Status code: 200
Name: zygarde-mega
Abilities: []


In [119]:
# Document the decision for Pokémon with empty abilities lists.
# The source API returns an empty list, so we preserve [] instead of replacing or removing the records.
empty_ability_count = (df["ability_count"] == 0).sum()
print(f"Pokémon with empty abilities lists: {empty_ability_count}")
print("Decision: Preserve empty abilities lists as [].")

Pokémon with empty abilities lists: 11
Decision: Preserve empty abilities lists as [].


In [120]:
# Check numeric columns for zero or negative values that may indicate invalid data.
# This helps us identify possible data-quality issues before transforming the dataset.
numeric_columns = ["height", "weight", "base_experience", "hp", "attack", "defense", "special_attack", "special_defense", "speed"]
for column in numeric_columns:
    print(f"\n{column}")
    print("Zero values:", (df[column] == 0).sum())
    print("Negative values:", (df[column] < 0).sum())


height
Zero values: 0
Negative values: 0

weight
Zero values: 1
Negative values: 0

base_experience
Zero values: 0
Negative values: 0

hp
Zero values: 0
Negative values: 0

attack
Zero values: 0
Negative values: 0

defense
Zero values: 0
Negative values: 0

special_attack
Zero values: 0
Negative values: 0

special_defense
Zero values: 0
Negative values: 0

speed
Zero values: 0
Negative values: 0


In [121]:
# Find the Pokémon with a weight of zero.
# We inspect the source record before deciding whether the zero is valid or an error.
zero_weight = df[df["weight"] == 0]
zero_weight[["id", "name", "height", "weight"]]

,id,name,height,weight
1214,10190,eternatus-eternamax,1000,0


In [122]:
# Check Eternatus-Eternamax directly from PokéAPI to verify whether weight = 0 comes from the source.
import requests
url = "https://pokeapi.co/api/v2/pokemon/eternatus-eternamax"
response = requests.get(url, timeout=20)
print("Status code:", response.status_code)
pokemon = response.json()
print("Name:", pokemon["name"])
print("Height:", pokemon["height"])
print("Weight:", pokemon["weight"])

Status code: 200
Name: eternatus-eternamax
Height: 1000
Weight: 0


In [123]:
# Document the decision for the zero weight value.
# PokéAPI itself returns weight = 0 for Eternatus-Eternamax, so we preserve the source value.
zero_weight_count = (df["weight"] == 0).sum()
print(f"Pokémon with weight = 0: {zero_weight_count}")
print("Decision: Preserve the source-provided weight value of 0.")

Pokémon with weight = 0: 1
Decision: Preserve the source-provided weight value of 0.


In [124]:
# Display the data types of all numeric columns.
# This helps us confirm that numeric fields are stored as numbers before transformation.
df[numeric_columns].dtypes

height               int64
weight               int64
base_experience    float64
hp                   int64
attack               int64
defense              int64
special_attack       int64
special_defense      int64
speed                int64
dtype: object

In [126]:
# Save the transformed Pokémon DataFrame as a clean CSV file for loading into PostgreSQL.
# This creates the dataset that will become our PostgreSQL source table.
clean_file = "pokemon_clean.csv"
df.to_csv(clean_file, index=False)
print(f"Saved cleaned dataset to: {clean_file}")

Saved cleaned dataset to: pokemon_clean.csv
